configfile: "config/config.yaml"

include: "rules/common.smk"
include: "rules/index.smk"
include: "rules/align.smk"
include: "rules/call.smk"

## input GFA from config['gfa']?
rule index_pangenome:
    input:
        expand('results/pg/{graph}.{ext}', graph=config['graph'], ext=['gbz', 'dist', 'min'])

## input from columns 'fq1' and 'fq2' of the TSV specified by config['sample_tsv'] (and for samples listed in config['sample'])
rule map_short_reads:
    input:
        expand('results/{sample}.{sample}.{graph}.gaf.gz', graph=config['graph'], sample=config['sample'])

## input from columns 'fq1' and 'fq2' of the TSV specified by config['sample_tsv'] (and for samples listed in config['sample'])
rule call_small_variants_from_short_reads:
    input:
        expand('results/{sample}/{sample}.{graph}.{realn}.snv_indels.vcf.gz', graph=config['graph'], sample=config['sample'], realn=config['bam_mode'])

rule genotype_variants_from_short_reads:
    input:
        expand('results/{sample}/{sample}.{graph}.gt.minlen{minlen}.vcf.gz', graph=config['graph'], sample=config['sample'], minlen=config['gt_min_var_len'])

rule call_structural_variants_from_short_reads:
    input:
        expand('results/{sample}/{sample}.{graph}.{realn}.sv_manta.vcf.gz', graph=config['graph'], sample=config['sample'], realn=config['bam_mode'])

rule prepare_unmapped_reads:
    input:
        expand('results/{sample}/{sample}.{graph}.unmapped.fq.gz', graph=config['graph'], sample=config['sample'])

rule all:
    input:
        expand('results/{sample}/fastq_qc.{sample}.zip.transferred', sample=config['sample']),
        expand('results/{sample}/{sample}.{graph}.unmapped.fq.gz.transferred', graph=config['graph'], sample=config['sample']),
        expand('results/{sample}/{sample}.{graph}.gaf.gz.transferred', graph=config['graph'], sample=config['sample']),
        expand('results/{sample}/{graph}.sample_pg.{sample}.gbz.transferred', graph=config['graph'], sample=config['sample']),
        expand('results/{sample}/{sample}.{graph}.{realn}.bam.transferred', graph=config['graph'], sample=config['sample'], realn=config['bam_mode']),
        expand('results/{sample}/{sample}.{graph}.{realn}.bam.bai.transferred', graph=config['graph'], sample=config['sample'], realn=config['bam_mode']),
        expand('results/{sample}/{sample}.{graph}.{realn}.snv_indels.vcf.gz.transferred', graph=config['graph'], sample=config['sample'], realn=config['bam_mode']),
        expand('results/{sample}/{sample}.{graph}.{realn}.snv_indels.g.vcf.gz.transferred', graph=config['graph'], sample=config['sample'], realn=config['bam_mode']),
        expand('results/{sample}/{sample}.{graph}.{realn}.sv_manta.vcf.gz.transferred', graph=config['graph'], sample=config['sample'], realn=config['bam_mode']),
        expand('results/{sample}/{sample}.{graph}.{realn}.sv_manta_candidates.vcf.gz.transferred', graph=config['graph'], sample=config['sample'], realn=config['bam_mode']),
        expand('results/{sample}/{sample}.{graph}.gt.minlen{minlen}.vcf.gz.transferred', graph=config['graph'], sample=config['sample'], minlen=config['gt_min_var_len']),
        expand('results/{sample}/{sample}.{graph}.{realn}.coverage.q.bed.gz.transferred', graph=config['graph'], sample=config['sample'], realn=config['bam_mode']),
        expand('results/{sample}/{sample}.{graph}.{realn}.coverage.bed.gz.transferred', graph=config['graph'], sample=config['sample'], realn=config['bam_mode'])

# transfer rules
rule save_to_local_storage:
    input: 'results/{sample}/{file}'
    output: 'results/{sample}/{file}.transferred'
    localrule: True
    priority: 100
    shell:
        """
        python3 transfer.run_genotoul.py -r upload -g {input} -l {input}
        """

rule get_fq:
    output: temp('fastqs/{file}')
    localrule: True
    shell:
        """
        python3 transfer.run_genotoul.py -r download -g {output} -l {output}
        """
